main/modes/games/2048/T48_logic.c

```diff
@@ -0,0 +1,372 @@
+/**
+ * @file T48_logic.c
+ * @author Jeremy Stintzcum (jeremy.stintzcum@gmail.com)
+ * @brief Game Logic for 2048
+ * @version 1.0.0
+ * @date 2024-08-16
+ *
+ * @copyright Copyright (c) 2024
+ *
+ */
+
+#include "T48_logic.h"
+
+/**
+ * @brief Sets an empty cell to 2 or 4 on 50/50 basis
+ *
+ * @param t48 Main Game Object
+ * @return int8_t Cell used
+ */
+static int8_t t48SetRandCell(t48_t* t48)
+{
+    int8_t cell;
+    do
+    {
+        cell = esp_random() % T48_BOARD_SIZE;
+    } while (t48->board[cell].val != 0);
+    int8_t rand = esp_random() % 10;
+    if (rand == 0)
+    {
+        t48->board[cell].val = 4; // 10%
+    }
+    else
+    {
+        t48->board[cell].val = 2; // 90%
+    }
+    t48->board[cell].state = NEW;
+    return cell;
+}
+
+/**
+ * @brief Updates the LEDs based on a direction and adds a new tile if valid
+ *
+ * @param t48 Main Game Object
+ * @param wasUpdated If the board has changed to a new config
+ * @param dir Direction used to update board
+ */
+static void t48BoardUpdate(t48_t* t48, bool wasUpdated, t48Direction_t dir)
+{
+    if (wasUpdated)
+    {
+        t48SetRandCell(t48);
+        led_t col = t48GetLEDColors(t48);
+        t48LightLEDs(t48, dir, col);
+        // FIXME: t48UpdateCells(t48);
+    }
+    else
+    {
+        led_t col = {.r = 200, .g = 200, .b = 200};
+        t48LightLEDs(t48, ALL, col);
+    }
+}
+
+/**
+ * @brief Merge a single row or column. Cells merged from CELL_SIZE-1 -> 0.
+ *
+ * @param t48 Main Game Object
+ * @param updated If board was already updated
+ */
+static void t48MergeSlice(t48_t* t48, bool* updated)
+{
+    for (uint8_t i = 0; i < T48_GRID_SIZE - 1; i++)
+    {
+        if (t48->sliceData.slice[i].val == 0)
+        {
+            continue;
+        }
+        if (t48->sliceData.slice[i].val == t48->sliceData.slice[i + 1].val)
+        {
+            *updated = true;
+            // Animate
+            t48->sliceData.dest[i + 1] = i; 
+            t48->sliceData.endVals[i] *= 2;
+            t48->sliceData.endVals[i + 1] = 0;
+
+            // Other
+            t48->sliceData.slice[i].state = MERGED;
+            // Math
+            t48->sliceData.slice[i].val *= 2;
+            t48->score += t48->sliceData.slice[i].val;
+            // Move if merged
+            for (uint8_t j = i + 1; j < T48_GRID_SIZE - 1; j++)
+            {
+                t48->sliceData.endVals[j] = t48->sliceData.endVals[j + 1];
+                t48->sliceData.slice[j].val = t48->sliceData.slice[j + 1].val;
+            }
+            // Add a 0 to end if merged
+            t48->sliceData.slice[T48_GRID_SIZE - 1].val   = 0;
+            t48->sliceData.slice[T48_GRID_SIZE - 1].state = STATIC;
+        }
+    }
+}
+
+void t48SlideDown(t48_t* t48)
```

Comment:
Intuition says the four `t48Slide*` functions could be merged into one, but I haven't looked too hard at the differences.